File: Part-1 BS on 1D Arrays/L1 Binary Search Introduction/CodeStudio.java
//* https://www.codingninjas.com/studio/problems/binary-search_972
//* https://youtu.be/MHf6awe89xw

public class Solution {
    public static int search(int []nums, int target) {
        //Your code goes here

        int low = 0, high = nums.length-1;

        while(low <= high){
            // int mid =  (high+low)/2;
            // int mid =  low + (high-low)/2;
            int mid =  high - (high-low)/2;

            if(nums[mid] == target){
                return mid;
            }
            else if(nums[mid] < target){
                // move right and eliminate left
                low = mid + 1;
            }
            else{
                // move left and eliminate right
                high = mid - 1;
            }
        }

        return -1;
    }

}

File: Part-1 BS on 1D Arrays/L4 Search Element in Rotated Sorted Array - I/CodeStudio.cpp
//* https://www.codingninjas.com/studio/problems/search-in-rotated-sorted-array_1082554
//* https://youtu.be/5qGrJbHhqFs

int search(vector<int>& arr, int n, int k)
{
    // Write your code here.
    // Return the position of K in ARR else return -1.

    int low = 0, high = n-1;

    while(low <= high){

        int mid = high - (high-low)/2;
        
        if(arr[mid] == k){
            return mid;
        }

        // if left half is sorted
        if(arr[low] <= arr[mid]){
            // is left contain the element k
            if(arr[low] <= k && k < arr[mid]){
                high = mid - 1;
            }
            else{
                // move to right
                low = mid + 1;
            }
        }
        // right is sorted 
        else{
            // is right contain the element k
            if (arr[mid] < k && k <= arr[high]){
                low = mid + 1;
            }
            else{
                // move left;
                high = mid - 1;
            }
        }
    }

    return -1;
}
